README.txt
==========================
 dataflake.ldapconnection
==========================

This package provides an abstraction layer on top of python-ldap. It
offers a connection object with simplified methods for inserting, 
modifying, searching and deleting records in the LDAP directory tree.
Failover/redundancy can be achieved by supplying connection data for 
more than one LDAP server.


Documentation
=============
Full documentation for the last released version is at
http://packages.python.org/dataflake.ldapconnection. For 
documentation matching the current SVN trunk please visit 
http://docs.dataflake.org/dataflake.ldapconnection.


Bug tracker
===========
A bug tracker is available at
https://bugs.launchpad.net/dataflake.ldapconnection



CHANGES.txt
Change log
==========

1.3 (unreleased)
----------------
- switched documentation to point to the new Git repository

- fakeldap: Add email characters and some non-ASCII characters to
  FLTR_RE, to be able to use them in searches.
  [stefan]

- fakeldap: Add unbind_s API.
  [stefan]

- fakeldap: Deep-copy entries before returning them.
  [stefan]

- fakeldap: Only return requested attributes.
  [stefan]

- fakeldap: Optionally disable password hashing.
  [stefan]

- fakeldap: Optionally maintain the 'memberOf' attribute of group
  members.
  [stefan]

- connection: Clean up encoding and decoding of values for when the
  api_encoding is set to None.
  [stefan]

- connection: New 'raw' kwarg for the search API. If true, search
  results are returned in the ldap_encoding.
  [stefan]

- connection: Store a connection's bind_dn and bind_pwd as is, and
  encode them before use.
  [stefan]

- connection: Fix a condition that caused rebinding to fail if only
  one of bind_dn and bind_pwd had changed.
  [stefan]

- connection: Add disconnect API.
  [stefan]

- connection: Allow to delete subsets from multi-valued attributes.
  [stefan]


1.2 (2010-08-09)
----------------
- Using ``id()`` is not random enough for a unique hash.


1.1 (2010-05-09)
----------------
- Updated Sphinx Makefile and configuration to be closer
  to the latest Sphinx version

- Greatly expand installation and testing documentation using
  ideas from Tres Seaver


1.0 (2010-04-12)
----------------
- Bug: fakeldap.FakeLDAPConnection wildcard searches did not 
  work correctly and returned too many matches.

- Bug: Improve behavior matching of standard python-ldap and 
  fakeldap by raising ldap.NO_SUCH_OBJECT where operations 
  target non-existing entries.

- Bug: Improve behavior matching of standard python-ldap and 
  fakeldap by raising ldap.ALREADY_EXISTS where operations 
  duplicate existing entries.

- Bug: Added tests for all fakeldap.FakeLDAPConnection methods
  and added tests for some other module classes and functions.

- Refactoring: Removed the fakeldap.initialize and explode_dn 
  functions. They were either not needed or needlessly 
  duplicating existing python-ldap features.

- Bug: python-ldap will no longer support the LDAP connection 
  class ldap.ldapobject.SmartLDAPObject with version 2.3.11.
  Replacing it with ReconnectLDAPObject.

- Bug: If a connection raised an LDAP exception inside 
  start_tls_s handling was broken.

- Feature: You can now add server definitions for servers that 
  support the StartTLS extended operation. Whereas the existing 
  secure connections using the ``ldaps`` protocol are encrypted 
  throughout, StartTLS is used through an unencrypted connection
  to request all further traffic to be encrypted.

- Refactoring: Switch tests to using the fakeldap LDAP connection 
  object wherever possible, and correct a few fakeldap and 
  LDAPConnection misbehaviors along the way.


1.0b1 (2010-02-01)
------------------

- Performing more rigorous input checking for DNs

- Made encoding/decoding more flexible by adding configuration 
  flags for the encoding used by the LDAP server and the 
  encoding for calls to and return values from the connection API.
  The default is backwards compatible (UTF-8 for the LDAP server 
  encoding, and Latin-15 for the API encoding).

- Factored the connection tests module into a series of modules,
  it was getting large and unwieldy.

- move the actual python-ldap connection from an attribute into
  a module-level cache since those connections cannot be
  pickled.

- Removed the rdn_attr attribute, which was used to try and 
  determine if a modify operation should trigger a modrdn. We 
  now fish the RDN attribute from the record's DN for this 
  purpose.

- Changed the way internal logging is done to avoid storing logger
  objects onto the connection instance unless it is explicitly
  specified. This means the instance is picklable when using the 
  default logging.

- Removed the `bind` method. There was no good reason to expose it 
  as part of the public API, and since bind operations are 
  re-done as part of all operations it would only serve to 
  confuse users.
  Users who want to use credentials other than the credentials 
  configured into the connection instance should pass them along 
  explicitly when invoking the operation.

- The search method now provides a default search subtree search 
  scope if none is specified.

- Creating a new instance does not require passing server data like 
  host, port and protocol anymore.

- replaced several methods with better alternatives from python-ldap,
  which also requires upping the dependency to python-ldap>=2.3.0, 
  and fixing up the tests.

- pare down fakeldap to not try and provide all kinds of constants 
  from python-ldap, but just a LDAP connection class.

- add a new method "bind" to rebind a connection, if the last bind
  differs from the desired bind.

- rename variable name "filter" with "fltr" to stop shadowing the 
  Python function "filter".

- added an interfaces file as documentation and "contract". This adds 
  a dependency on zope.interface.

- removed unused argument "login_attr" from constructor argument list

- LDAPConnection objects now accept more than a single server definition.
  Failover between connections is triggered by connection or operation
  timeouts. Added API to add and remove server definitions at runtime.

- all those methods causing LDAP operations to be performed accept 
  optional bind_dn and bind_pwd named arguments to rebind with the 
  provided credentials instead of those credentials stored in the 
  LDAPConnection instance. This represents an API change for the 
  `insert`, `modify` and `delete` methods.


0.4 (2008-12-25)
----------------
- fakeldap bug: the modify_s method would expect changes of type MOD_DELETE
  to come with a list of specific attribute values to delete. Now the 

dataflake/ldapconnection/tests/base.py
##############################################################################
#
# Copyright (c) 2008-2009 Jens Vagelpohl and Contributors. All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
""" unit tests base classes

$Id$
"""

import unittest

from dataflake.ldapconnection.connection import connection_cache
from dataflake.ldapconnection.tests import fakeldap

class LDAPConnectionTests(unittest.TestCase):

    def setUp(self):
        super(LDAPConnectionTests, self).setUp()
        # Put a record into the tree
        fakeldap.addTreeItems('dc=localhost')

    def tearDown(self):
        super(LDAPConnectionTests, self).tearDown()
        fakeldap.clearTree()
        connection_cache.invalidate()

    def _getTargetClass(self):
        from dataflake.ldapconnection.connection import LDAPConnection
        return LDAPConnection

    def _makeOne(self, *args, **kw):
        conn = self._getTargetClass()(*args, **kw)
        conn.api_encoding = 'iso-8859-1'
        conn.ldap_encoding = 'UTF-8'
        return conn

    def _makeSimple(self):
        conn = self._makeOne('host', 636, 'ldap', fakeldap.FakeLDAPConnection)
        conn.api_encoding = 'iso-8859-1'
        conn.ldap_encoding = 'UTF-8'
        return conn

    def _makeRaising(self, raise_on, exc_class, exc_arg=None):
        ldap_connection = fakeldap.RaisingFakeLDAPConnection('conn_string')
        ldap_connection.setExceptionAndMethod(raise_on, exc_class, exc_arg)
        def factory(conn_string):
            ldap_connection.conn_string = conn_string
            return ldap_connection
        conn = self._makeOne('host', 389, 'ldaptls', factory)

        return conn, ldap_connection

    def _makeFixedResultConnection(self, results):
        ldap_connection = fakeldap.FixedResultFakeLDAPConnection()
        ldap_connection.search_results = results
        def factory(conn_string):
            ldap_connection.conn_string = conn_string
            return ldap_connection
        conn = self._makeOne('host', 389, 'ldaptls', factory)

        return conn

    def _factory(self, connection_string):
        of = fakeldap.FakeLDAPConnection(connection_string)
        return of

    def _addRecord(self, dn, **kw):
        record = fakeldap.addTreeItems(dn)
        for key, value in kw.items():
            if key.lower() == 'userpassword':
                value = [fakeldap.hash_pwd(value)]
            elif isinstance(value, basestring):
                value = [value]
            record[key] = value


class FakeLDAPTests(unittest.TestCase):

    def setUp(self):
        from dataflake.ldapconnection.tests import fakeldap
        fakeldap.addTreeItems('ou=users,dc=localhost')
        fakeldap.addTreeItems('ou=groups,dc=localhost')

    def tearDown(self):
        from dataflake.ldapconnection.tests import fakeldap
        fakeldap.clearTree()

    def _getTargetClass(self):
        from dataflake.ldapconnection.tests.fakeldap import FakeLDAPConnection
        return FakeLDAPConnection

    def _makeOne(self, *args, **kw):
        conn = self._getTargetClass()(*args, **kw)
        return conn

    def _addUser(self, name, mail=None):
        conn = self._makeOne()
        user_dn = 'cn=%s,ou=users,dc=localhost' % name
        user_pwd = '%s_secret' % name

        if conn.hash_password:
            pwd = fakeldap.hash_pwd(user_pwd)
        else:
            pwd = user_pwd

        user = [ ('cn', [name])
               , ('userPassword', [pwd])
               , ('objectClass', ['top', 'person'])
               ]
        if mail is not None:
            user.append(('mail', [mail]))

        conn.add_s(user_dn, user)
        return (user_dn, user_pwd)

    def _addGroup(self, name, members=None):
        conn = self._makeOne()
        group_dn = 'cn=%s,ou=groups,dc=localhost' % name

        group = [ ('cn', [name])
                , ('objectClass', ['top', 'group'])
                ]
        if members is not None:
            members = ['cn=%s,ou=users,dc=localhost' % x for x in members]
            group.append((conn.member_attr, members))

        conn.add_s(group_dn, group)
        return group_dn



dataflake/ldapconnection/tests/dummy.py
##############################################################################
#
# Copyright (c) 2008-2010 Jens Vagelpohl and Contributors. All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
""" dummy: dummy test fixtures

$Id$
"""

ISO_8859_1_UNICODE = u'\xe4\xf6\xfc\xdf' # Umlauts a, o, u, sharp s
ISO_8859_1_UTF8 = ISO_8859_1_UNICODE.encode('UTF-8')
ISO_8859_1_ENCODED = ISO_8859_1_UNICODE.encode('iso-8859-1')
ISO_8859_7_UNICODE = u'\u03b1\u03b2\u03b3\u03b4' # Greek alpha beta gamma delta
ISO_8859_7_UTF8 = ISO_8859_7_UNICODE.encode('UTF-8')
ISO_8859_7_ENCODED = ISO_8859_7_UNICODE.encode('iso-8859-7')


bootstrap.py
##############################################################################
#
# Copyright (c) 2006 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""Bootstrap a buildout-based project

Simply run this script in a directory containing a buildout.cfg.
The script accepts buildout command-line options, so you can
use the -c option to specify an alternate configuration file.

$Id: bootstrap.py 1914 2010-04-12 11:37:04Z jens $
"""

import os, shutil, sys, tempfile, urllib2
from optparse import OptionParser

tmpeggs = tempfile.mkdtemp()

is_jython = sys.platform.startswith('java')

# parsing arguments
parser = OptionParser()
parser.add_option("-v", "--version", dest="version",
                          help="use a specific zc.buildout version")
parser.add_option("-d", "--distribute",
                   action="store_true", dest="distribute", default=False,
                   help="Use Distribute rather than Setuptools.")

parser.add_option("-c", None, action="store", dest="config_file",
                   help=("Specify the path to the buildout configuration "
                         "file to be used."))

options, args = parser.parse_args()

# if -c was provided, we push it back into args for buildout' main function
if options.config_file is not None:
    args += ['-c', options.config_file]

if options.version is not None:
    VERSION = '==%s' % options.version
else:
    VERSION = ''

USE_DISTRIBUTE = options.distribute
args = args + ['bootstrap']

to_reload = False
try:
    import pkg_resources
    if not hasattr(pkg_resources, '_distribute'):
        to_reload = True
        raise ImportError
except ImportError:
    ez = {}
    if USE_DISTRIBUTE:
        exec urllib2.urlopen('http://python-distribute.org/distribute_setup.py'
                         ).read() in ez
        ez['use_setuptools'](to_dir=tmpeggs, download_delay=0, no_fake=True)
    else:
        exec urllib2.urlopen('http://peak.telecommunity.com/dist/ez_setup.py'
                             ).read() in ez
        ez['use_setuptools'](to_dir=tmpeggs, download_delay=0)

    if to_reload:
        reload(pkg_resources)
    else:
        import pkg_resources

if sys.platform == 'win32':
    def quote(c):
        if ' ' in c:
            return '"%s"' % c # work around spawn lamosity on windows
        else:
            return c
else:
    def quote (c):
        return c

cmd = 'from setuptools.command.easy_install import main; main()'
ws  = pkg_resources.working_set

if USE_DISTRIBUTE:
    requirement = 'distribute'
else:
    requirement = 'setuptools'

if is_jython:
    import subprocess

    assert subprocess.Popen([sys.executable] + ['-c', quote(cmd), '-mqNxd',
           quote(tmpeggs), 'zc.buildout' + VERSION],
           env=dict(os.environ,
               PYTHONPATH=
               ws.find(pkg_resources.Requirement.parse(requirement)).location
               ),
           ).wait() == 0

else:
    assert os.spawnle(
        os.P_WAIT, sys.executable, quote (sys.executable),
        '-c', quote (cmd), '-mqNxd', quote (tmpeggs), 'zc.buildout' + VERSION,
        dict(os.environ,
            PYTHONPATH=
            ws.find(pkg_resources.Requirement.parse(requirement)).location
            ),
        ) == 0

ws.add_entry(tmpeggs)
ws.require('zc.buildout' + VERSION)
import zc.buildout.buildout
zc.buildout.buildout.main(args)
shutil.rmtree(tmpeggs)


dataflake/ldapconnection/utils.py
##############################################################################
#
# Copyright (c) 2008-2010 Jens Vagelpohl and Contributors. All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
""" Utility functions and constants

$Id: utils.py 1485 2008-06-04 16:08:38Z jens $
"""

import ldap

BINARY_ATTRIBUTES = ('objectguid', 'jpegphoto')

def escape_dn(dn):
    """ Escape all characters that need escaping for a DN, see RFC 2253 
    """
    if dn is None:
        return None

    return ldap.dn.dn2str(ldap.dn.str2dn(dn))



dataflake/ldapconnection/connection.py
##############################################################################
#
# Copyright (c) 2008-2010 Jens Vagelpohl and Contributors. All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
""" LDAPConnection: A class modeling a LDAP server connection

Instances of this class offer a simplified API to do searches, insertions, 
deletions or modifications.

$Id$
"""

import codecs
import ldap
from ldap.dn import explode_dn
from ldap.dn import dn2str
from ldap.dn import str2dn
from ldap.filter import filter_format
from ldap.ldapobject import ReconnectLDAPObject
import ldapurl
import logging
from random import random

from zope.interface import implements

from dataflake.cache.simple import LockingSimpleCache
from dataflake.ldapconnection.interfaces import ILDAPConnection
from dataflake.ldapconnection.utils import BINARY_ATTRIBUTES
from dataflake.ldapconnection.utils import escape_dn

default_logger = logging.getLogger('dataflake.ldapconnection')
connection_cache = LockingSimpleCache()
_marker = ()


class LDAPConnection(object):
    """ LDAPConnection object

    See `interfaces.py` for interface documentation.
    """

    implements(ILDAPConnection)

    def __init__( self, host='', port=389, protocol='ldap'
                , c_factory=ReconnectLDAPObject, rdn_attr='', bind_dn=''
                , bind_pwd='', read_only=False
                , conn_timeout=-1, op_timeout=-1, logger=None
                ):
        """ LDAPConnection initialization
        """
        # Empty values here mean "use unicode"
        self.ldap_encoding = 'UTF-8'
        self.api_encoding = 'iso-8859-15'

        self.bind_dn = bind_dn
        self.bind_pwd = bind_pwd
        self.read_only = read_only
        self.c_factory = c_factory
        self._logger = logger
        self.hash = id(self) + random()

        self.servers = {}
        if host:
            self.addServer(host, port, protocol, conn_timeout, op_timeout)

    def logger(self):
        """ Get the logger
        """
        if self._logger is None:
            return default_logger

        return self._logger

    def addServer(self, host, port, protocol, conn_timeout=-1, op_timeout=-1):
        """ Add a server definition to the list of servers used
        """
        start_tls = False
        if protocol == 'ldaptls':
            protocol = 'ldap'
            start_tls = True
        l = ldapurl.LDAPUrl(urlscheme=protocol, hostport='%s:%s' % (host, port))
        server_url = l.initializeUrl()
        self.servers[server_url] = { 'url' : server_url
                                   , 'conn_timeout' : conn_timeout
                                   , 'op_timeout' : op_timeout
                                   , 'start_tls': start_tls
                                   }

    def removeServer(self, host, port, protocol):
        """ Remove a server definition from the list of servers used
        """
        l = ldapurl.LDAPUrl(urlscheme=protocol, hostport='%s:%s' % (host, port))
        server_url = l.initializeUrl()
        if server_url in self.servers.keys():
            del self.servers[server_url]

    def connect(self, bind_dn=None, bind_pwd=None):
        """ initialize an ldap server connection 

        This method returns an instance of the underlying `python-ldap` 
        connection class. It does not need to be called explicitly, all
        other operations call it implicitly.
        """
        if len(self.servers.keys()) == 0:
            raise RuntimeError('No servers defined')

        if bind_dn is None:
            bind_dn = escape_dn(self._encode_incoming(self.bind_dn))
            bind_pwd = self._encode_incoming(self.bind_pwd)
        else:
            bind_dn = escape_dn(self._encode_incoming(bind_dn))
            bind_pwd = self._encode_incoming(bind_pwd)

        conn = self._getConnection()
        if conn is None:
            for server in self.servers.values():
                try:
                    conn = self._connect( server['url']
                                        , conn_timeout=server['conn_timeout']
                                        , op_timeout=server['op_timeout']
                                        )
                    if server.get('start_tls', None):
                        conn.start_tls_s()
                    break
                except (ldap.SERVER_DOWN, ldap.TIMEOUT, ldap.LOCAL_ERROR), e:
                    conn = None
                    continue

            if conn is None:
                exception_string = str(e or 'no exception')
                msg = 'Failure connecting, last attempt: %s (%s)' % (
                            server['url'], str(e or 'no exception'))
                self.logger().critical(msg, exc_info=1)

                if e:
                    raise e

            connection_cache.set(self.hash, conn)

        last_bind = getattr(conn, '_last_bind', None)
        if ( not last_bind or
             last_bind[1][0] != bind_dn or
             last_bind[1][1] != bind_pwd ):
            conn.simple_bind_s(bind_dn, bind_pwd)

        return conn

    def _getConnection(self):
        """ Private helper to get my connection out of the cache
        """
        return connection_cache.get(self.hash)

    def _connect( self
                , connection_string
                , conn_timeout=5
                , op_timeout=-1
                ):
        """ Factored out to allow usage by other pieces 

        user_dn is assumed to have been encoded/escaped correctly
        """
        connection = self.c_factory(connection_string)

        # Deny auto-chasing of referrals to be safe, we handle them instead
        try:
            connection.set_option(ldap.OPT_REFERRALS, ldap.DEREF_NEVER)
        except ldap.LDAPError: # Cannot set referrals, so do nothing
            pass

        # Set the connection timeout
        if conn_timeout > 0:
            connection.set_option(ldap.OPT_NETWORK_TIMEOUT, conn_timeout)

        # Set the operations timeout
        if op_timeout > 0:
            connection.timeout = op_timeout

        return connection

    def disconnect(self):
        """ Unbind the connection and invalidate the cache
        """
        conn = self._getConnection()
        if conn is not None:
            connection_cache.invalidate(self.hash)
            conn.unbind_s()

    def search( self
              , base
              , scope=ldap.SCOPE_SUBTREE
              , fltr='(objectClass=*)'
              , attrs=None
              , convert_filter=True
              , bind_dn=None
              , bind_pwd=None
              , raw=False
              ):
        """ Search for entries in the database
        """
        result = {'size': 0, 'results': [], 'exception': ''}
        if convert_filter:
            fltr = self._encode_incoming(fltr)
        base = escape_dn(self._encode_incoming(base))
        connection = self.connect(bind_dn=bind_dn, bind_pwd=bind_pwd)

        try:
            res = connection.search_s(base, scope, fltr, attrs)
        except ldap.PARTIAL_RESULTS:
            res_type, res = connection.result(all=0)
        except ldap.REFERRAL, e:
            connection = self._handle_referral(e)

            try:
                res = connection.search_s(base, scope, fltr, attrs)
            except ldap.PARTIAL_RESULTS:
                res_type, res = connection.result(all=0)

        for rec_dn, rec_dict in res:
            # When used against Active Directory, "rec_dict" may not be
            # be a dictionary in some cases (instead, it can be a list)
            # An example of a useless "res" entry that can be ignored
            # from AD is
            # (None, ['ldap://ForestDnsZones.PORTAL.LOCAL/DC=ForestDnsZones,DC=PORTAL,DC=LOCAL'])
            # This appears to be some sort of internal referral, but
            # we can't handle it, so we need to skip over it.
            try:
                items =  rec_dict.items()
            except AttributeError:
                # 'items' not found on rec_dict
                continue

            if raw:
                rec_dict['dn'] = rec_dn
            else:
                for key, value in items:
                    if key.lower() not in BINARY_ATTRIBUTES:
                        if not isinstance(value, basestring):
                            for i in range(len(value)):
                                value[i] = self._encode_outgoing(value[i])
                        else:
                            rec_dict[key] = self._encode_outgoing(value)
                rec_dict['dn'] = self._encode_outgoing(rec_dn)

            result['results'].append(rec_dict)
            result['size'] += 1

        return result

    def insert(self, base, rdn, attrs=None, bind_dn=None, bind_pwd=None):
        """ Insert a new record 

        attrs is expected to be a mapping where the value may be a string
        or a sequence of strings. 
        Multiple values may be expressed as a single string if the values 
        are semicolon-delimited.
        Values can be marked as binary values, meaning they are not encoded
        as UTF-8, by appending ';binary' to the key.
        """
        self._complainIfReadOnly()
        base = escape_dn(self._encode_incoming(base))
        rdn = escape_dn(self._encode_incoming(rdn))

        dn = rdn + ',' + base
        attribute_list = []
        attrs = attrs and attrs or {}

        for attr_key, values in attrs.items():
            if attr_key.endswith(';binary'):
                is_binary = True
                attr_key = attr_key[:-7]
            else:
                is_binary = False

            if isinstance(values, basestring) and not is_binary:
                values = [x.strip() for x in values.split(';')]

            if values != ['']:
                if not is_binary:
                    values = [self._encode_incoming(x) for x in values]
                attribute_list.append((attr_key, values))

        try:
            connection = self.connect(bind_dn=bind_dn, bind_pwd=bind_pwd)
            connection.add_s(dn, attribute_list)
        except ldap.REFERRAL, e:
            connection = self._handle_referral(e)
            connection.add_s(dn, attribute_list)

    def delete(self, dn, bind_dn=None, bind_pwd=None):
        """ Delete a record 
        """
        self._complainIfReadOnly()

        dn = escape_dn(self._encode_incoming(dn))

        try:
            connection = self.connect(bind_dn=bind_dn, bind_pwd=bind_pwd)
            connection.delete_s(dn)
        except ldap.REFERRAL, e:
            connection = self._handle_referral(e)
            connection.delete_s(dn)

    def modify(self, dn, mod_type=None, attrs=None, bind_dn=None, bind_pwd=None):
        """ Modify a record 
        """
        self._complainIfReadOnly()

        unescaped_dn = self._encode_incoming(dn)
        dn = escape_dn(unescaped_dn)
        res = self.search( base=unescaped_dn
                         , scope=ldap.SCOPE_BASE
                         , bind_dn=bind_dn
                         , bind_pwd=bind_pwd
                         , raw=True
                         )
        attrs = attrs and attrs or {}
        cur_rec = res['results'][0]
        mod_list = []

        for key, values in attrs.items():

            if key.endswith(';binary'):
                key = key[:-7]
            elif isinstance(values, basestring):
                values = [self._encode_incoming(x) for x in values.split(';')]
            else:
                values = [self._encode_incoming(x) for x in values]

            if mod_type is None:
                if not cur_rec.has_key(key) and values != ['']:
                    mod_list.append((ldap.MOD_ADD, key, values))
                elif cur_rec.get(key,['']) != values and values not in ([''],[]):
                    mod_list.append((ldap.MOD_REPLACE, key, values))
                elif cur_rec.has_key(key) and values in ([''], []):
                    mod_list.append((ldap.MOD_DELETE, key, None))
            elif mod_type in (ldap.MOD_ADD, ldap.MOD_DELETE) and values == ['']:
                continue
            elif ( mod_type == ldap.MOD_DELETE and
                   set(values).difference(set(cur_rec.get(key, []))) ):
                continue
            else:
                mod_list.append((mod_type, key, values))

        try:
            connection = self.connect(bind_dn=bind_dn, bind_pwd=bind_pwd)

            dn_parts = str2dn(dn)
            rdn = dn_parts[0]
            rdn_attr = rdn[0][0]
            raw_rdn = attrs.get(rdn_attr, '')
            if isinstance(raw_rdn, basestring):
                raw_rdn = [raw_rdn]
            new_rdn = raw_rdn[0]

            if new_rdn:
                rdn_value = self._encode_incoming(new_rdn)
                if rdn_value != cur_rec.get(rdn_attr)[0]:
                    dn_parts[0] = [(rdn_attr, rdn_value, 1)]
                    raw_utf8_rdn = rdn_attr + '=' + rdn_value
                    new_rdn = escape_dn(raw_utf8_rdn)
                    connection.modrdn_s(dn, new_rdn)
                    dn = dn2str(dn_parts)

            if mod_list:
                connection.modify_s(dn, mod_list)
            else:
                debug_msg = 'Nothing to modify: %s' % dn
                self.logger().debug(debug_msg)

        except ldap.REFERRAL, e:
            connection = self._handle_referral(e)
            connection.modify_s(dn, mod_list)

    def _handle_referral(self, exception):
        """ Handle a referral specified in the passed-in exception 
        """
        payload = exception.args[0]
        info = payload.get('info')
        ldap_url = info[info.find('ldap'):]

        if ldapurl.isLDAPUrl(ldap_url):
            conn_str = ldapurl.LDAPUrl(ldap_url).initializeUrl()
            conn = self._connect(conn_str)
            conn.simple_bind_s(self._encode_incoming(self.bind_dn),
                               self._encode_incoming(self.bind_pwd))
            return conn
        else:
            raise ldap.CONNECT_ERROR, 'Bad referral "%s"' % str(exception)

    def _complainIfReadOnly(self):
        """ Raise RuntimeError if the connection is set to `read-only`

        This method should be called before any directory tree modfication
        """
        if self.read_only:
            raise RuntimeError(
                'Running in read-only mode, directory modifications disabled')

    def _encode_incoming(self, value):
        """ Encode a string value to the encoding set for the LDAP server

        - if "value" is unicode, it will be encoded to self.ldap_encoding, but
          only if self.ldap_encoding is set.
        - if "value" is not unicode, it is assumed to be encoded as 
          self.api_encoding. It is decoded and encoded to self.ldap_encoding
          if self.ldap_encoding is set, unless self.api_encoding and 
          self.ldap_encoding are identical. In that case the passed-in value 
          is handed back unchanged.
        """
        if value is None:
            return None

        elif isinstance(value, unicode):
            if self.ldap_encoding:
                value = value.encode(self.ldap_encoding)

        else:
            if self.api_encoding != self.ldap_encoding:
                if self.api_encoding:
                    value = value.decode(self.api_encoding)

                    if self.ldap_encoding:
                        value = value.encode(self.ldap_encoding)
        
        return value

    def _encode_outgoing(self, value):
        """ Encode a string value to the API encoding

        - if "value" is unicode, it will be encoded to self.api_encoding, but
          only if self.api_encoding is set.
        - if "value" is not unicode, it is assumed to be encoded as 
          self.ldap_encoding. It is decoded and encoded to self.api_encoding
          if self.api_encoding is set, unless self.ldap_encoding and 
          self.api_encoding are identical. In that case the passed-in value 
          is handed back unchanged.
        """
        if value is None:
            return None

        elif isinstance(value, unicode):
            if self.api_encoding:
                value = value.encode(self.api_encoding)

        else:
            if self.api_encoding != self.ldap_encoding:
                if self.ldap_encoding:
                    value = value.decode(self.ldap_encoding)

                    if self.api_encoding:
                        value = value.encode(self.api_encoding)
        
        return value
